Issues with GNOME 3 and Nvidia EGLStream

> As long as the PR was tested on these commits with this config
> * https://discourse.nixos.org/t/need-gnome-3-tester-for-nvidia-eglstream-support/2904/2?u=worldofpeace

Yes.

> I don't think the firefox package in nixpkgs supports wayland yet.

We have a `firefox-wayland` package which is what I tried, but since it’s a thin wrapper around `firefox` it’s possible that it’s still not built with the right options.

Also I’ve noticed these messages from `journalctl --user -b 0 -t gnome-shell`:

```
gnome-shell[2106]: xwayland glamor: GBM backend (default) is not available
gnome-shell[2106]: Missing Wayland requirements for glamor GBM backend
gnome-shell[2106]: glamor: 'wl_eglstream_display' not supported
gnome-shell[2106]: Missing Wayland requirements for glamor EGLStream backend
gnome-shell[2106]: Failed to initialize glamor, falling back to sw
```

`xdpyinfo` does not show the `GLX` extension, and GLX applications fail to start.

_Originally posted by @andersk in https://github.com/NixOS/nixpkgs/pull/70295#issuecomment-558936554_

I probably merged https://github.com/NixOS/nixpkgs/pull/70295, too early.
I'm not totally sure this is working properly.

I've updated to the latest nixos-unstable (45fc7d4a35c5) which includes updated `xwayland` & `egl-wayland`, but I'm still facing issues.

While it's generally working I have problems with some apps.
`kitty`:
```
[192 13:26:15.149439] [glfw error 65542]: EGL: Failed to get EGL display: Success
[192 13:26:15.159831] Traceback (most recent call last):
  File "/nix/store/sdr311kij8ipyyg0was3a1mik84vmjl6-kitty-0.21.2/bin/../lib/kitty/kitty/main.py", line 344, in main
    _main()
  File "/nix/store/sdr311kij8ipyyg0was3a1mik84vmjl6-kitty-0.21.2/bin/../lib/kitty/kitty/main.py", line 337, in _main
    run_app(opts, cli_opts, bad_lines)
  File "/nix/store/sdr311kij8ipyyg0was3a1mik84vmjl6-kitty-0.21.2/bin/../lib/kitty/kitty/main.py", line 183, in __call__
    _run_app(opts, args, bad_lines)
  File "/nix/store/sdr311kij8ipyyg0was3a1mik84vmjl6-kitty-0.21.2/bin/../lib/kitty/kitty/main.py", line 156, in _run_app
    window_id = create_os_window(
ValueError: Failed to create GLFWwindow
```
`alacritty`:
```
Created log file at "/run/user/1000/Alacritty-8434.log"
[2021-07-11 13:27:51.630725226] [ERROR] [alacritty] Alacritty encountered an unrecoverable error:
                                                    
                                                        Error creating GL context; Could not create EGL display objec
```
`glxinfo`:
```
name of display: :0
Error: couldn't find RGB GLX visual or fbconfi
```

Additionally in my system logs I can see:
```
Jul 11 13:16:17 felix-nixos gnome-shell[2720]: Failed to initialize glamor, falling back to sw
Jul 11 13:16:17 felix-nixos gnome-shell[2720]: Missing Wayland requirements for glamor EGLStream backend
Jul 11 13:16:17 felix-nixos gnome-shell[2720]: glamor: 'wl_eglstream_display' not supported
Jul 11 13:16:17 felix-nixos gnome-shell[2720]: Missing Wayland requirements for glamor GBM backend
```

`firefox` & `firefox-wayland` are both working.

My relevant config:
```
  services.xserver.videoDrivers = [ "nvidia" ];
  hardware.nvidia.modesetting.enable = true;
  hardware.nvidia.package = config.boot.kernelPackages.nvidiaPackages.beta;
  programs.xwayland.enable = true;

  services.xserver.enable = true;
  services.xserver.displayManager.gdm.enable = true;
  services.xserver.displayManager.gdm.nvidiaWayland = true;
  services.xserver.displayManager.gnome.enable = true;
```

I also tried using an overlay to apply the new meson flags from @baracoder's overlay:
```
  xwayland = super.xwayland.overrideAttrs (a: rec {
    mesonFlags = a.mesonFlags ++ [ "-Ddri3=true" "-Dglamor=true" ];
  });
```

Is there anything else I need to set up?

I found this issue because I had the same problem as @felschr.

After reading @babariviere's [comment](https://github.com/NixOS/nixpkgs/issues/75131#issuecomment-614621092), I tried to create a different `/etc/egl/egl_external_platform.d/nvidia_wayland.json` file (listed in the end). And that solved my problem. With this file, I can open steam. However, alacritty still has an error. Maybe that relates to another [issue](https://github.com/nvidia/egl-wayland/issues/34).
```
Created log file at "/tmp/Alacritty-4532.log"
[2021-08-24 20:20:17.028947543] [ERROR] [alacritty] Alacritty encountered an unrecoverable error:
                                                    
                                                    	Error creating GL context; eglCreateWindowSurface failed
```

The cause of this problem is that even though both `egl-wayland` and `nvidia-x11` packages provide `libnvidia-egl-wayland.so`, `libnvidia-egl-wayland.so` cannot be found without a json file. The [nvidia egl-wayland repo](https://github.com/NVIDIA/egl-wayland/#building-and-installing-the-library) says:
```
Notes:

The NVIDIA EGL driver uses a JSON-based loader to load all EGL External platforms available on the system.

If this library is not installed as part of a NVIDIA driver installation, a JSON configuration file must be manually added in order to make the library work with the NVIDIA driver.

The default EGL External platform JSON configuration directory is:

/usr/share/egl/egl_external_platform.d/
```

So, maybe we should add a json file `/usr/share/egl/egl_external_platform.d
/xxx.json` to the package of `nvidia-x11` and `egl-wayland`. I have tried to use a json file to load `libnvidia-egl-wayland.so` from each of them, and both of them work. The only difference I find is that using `libnvidia-egl-wayland.so` from `egl-wayland` helps gnome setting recognizes my nvidia card, while using `nvidia-x11`, gnome setting says my video card is "Unknown".

In the end, I will post my related config:
```
services.xserver = {
  enable = true;
  videoDrivers = [ "nvidia" ];
  displayManager.gdm = {
    enable = true;
    nvidiaWayland = true;
  };
  desktopManager.gnome.enable = true;
};
programs.xwayland.enable = true;
hardware.nvidia.modesetting.enable = true;
```
/etc/egl/egl_external_platform.d/nvidia_wayland.json
```
{
    "file_format_version" : "1.0.0",
    "ICD" : {
        "library_path" : "/run/opengl-driver/lib/libnvidia-egl-wayland.so"
    }
}
```

--- a/nixos/modules/hardware/video/nvidia.nix
+++ b/nixos/modules/hardware/video/nvidia.nix
@@ -284,6 +284,10 @@ in
       source = "${nvidia_x11.bin}/share/nvidia/nvidia-application-profiles-rc";
     };
 
+    # 'nvidia_x11' installs it's files to /run/opengl-driver/...
+    environment.etc."egl/egl_external_platform.d".source =
+      "/run/opengl-driver/share/egl/egl_external_platform.d/";
+
     hardware.opengl.package = mkIf (!offloadCfg.enable) nvidia_x11.out;
     hardware.opengl.package32 = mkIf (!offloadCfg.enable) nvidia_x11.lib32;
     hardware.opengl.extraPackages = optional offloadCfg.enable nvidia_x11.out;

--- a/pkgs/os-specific/linux/nvidia-x11/builder.sh
+++ b/pkgs/os-specific/linux/nvidia-x11/builder.sh
@@ -108,6 +108,14 @@ installPhase() {
 
             install -Dm644 10_nvidia.json.fixed $i/share/glvnd/egl_vendor.d/10_nvidia.json
             install -Dm644 10_nvidia_wayland.json.fixed $i/share/egl/egl_external_platform.d/10_nvidia_wayland.json
+
+            if [[ -f "15_nvidia_gbm.json" ]]; then
+              sed -E "s#(libnvidia-egl-gbm)#$i/lib/\\1#" 15_nvidia_gbm.json > 15_nvidia_gbm.json.fixed
+              install -Dm644 15_nvidia_gbm.json.fixed $i/share/egl/egl_external_platform.d/15_nvidia_gbm.json
+
+              mkdir -p $i/lib/gbm
+              ln -s $i/lib/libnvidia-allocator.so $i/lib/gbm/nvidia-drm_gbm.so
+            fi
         fi
 
     done

--- a/pkgs/os-specific/linux/nvidia-x11/builder.sh
+++ b/pkgs/os-specific/linux/nvidia-x11/builder.sh
@@ -106,8 +106,8 @@ installPhase() {
             sed -E "s#(libEGL_nvidia)#$i/lib/\\1#" 10_nvidia.json > 10_nvidia.json.fixed
             sed -E "s#(libnvidia-egl-wayland)#$i/lib/\\1#" 10_nvidia_wayland.json > 10_nvidia_wayland.json.fixed
 
-            install -Dm644 10_nvidia.json.fixed $i/share/glvnd/egl_vendor.d/nvidia.json
-            install -Dm644 10_nvidia_wayland.json.fixed $i/share/glvnd/egl_vendor.d/nvidia_wayland.json
+            install -Dm644 10_nvidia.json.fixed $i/share/glvnd/egl_vendor.d/10_nvidia.json
+            install -Dm644 10_nvidia_wayland.json.fixed $i/share/egl/egl_external_platform.d/10_nvidia_wayland.json
         fi
 
     done

--- a/pkgs/os-specific/linux/nvidia-x11/generic.nix
+++ b/pkgs/os-specific/linux/nvidia-x11/generic.nix
@@ -41,7 +41,7 @@ let
   i686bundled = versionAtLeast version "391" && !disable32Bit;
 
   libPathFor = pkgs: pkgs.lib.makeLibraryPath [ pkgs.libdrm pkgs.xorg.libXext pkgs.xorg.libX11
-    pkgs.xorg.libXv pkgs.xorg.libXrandr pkgs.xorg.libxcb pkgs.zlib pkgs.stdenv.cc.cc ];
+    pkgs.xorg.libXv pkgs.xorg.libXrandr pkgs.xorg.libxcb pkgs.zlib pkgs.stdenv.cc.cc pkgs.wayland ];
 
   self = stdenv.mkDerivation {
     name = "nvidia-x11-${version}${nameSuffix}";

